idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,850 @@
+            for t, p in zip(time_index, price_forecasts[i]):
+                self.model.fs[i].energy_price[t] = p
+
+        return
+
+    def _assemble_bids(self):
+
+        """
+        This methods extract the bids out of the stochastic programming model and
+        organize them.
+
+        Arguments:
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure. {t: {gen:{power: cost}}}.
+        """
+
+        bids = {}
+        gen = self.generator
+
+        for i in self.model.SCENARIOS:
+            time_index = self.model.fs[i].energy_price.index_set()
+            for t in time_index:
+
+                if t not in bids:
+                    bids[t] = {}
+                if gen not in bids[t]:
+                    bids[t][gen] = {}
+
+                power = round(pyo.value(self.model.fs[i].power_output_ref[t]), 2)
+                marginal_cost = round(pyo.value(self.model.fs[i].energy_price[t]), 2)
+
+                if power < self.bidding_model_object.pmin:
+                    continue
+                elif power in bids[t][gen]:
+                    bids[t][gen][power] = min(bids[t][gen][power], marginal_cost)
+                else:
+                    bids[t][gen][power] = marginal_cost
+
+        for t in time_index:
+
+            # make sure the orignal points in the bids
+            for power, marginal_cost in self.bidding_model_object.default_bids.items():
+                if power not in bids[t][gen]:
+                    bids[t][gen][power] = marginal_cost
+
+            pmin = self.bidding_model_object.pmin
+
+            # sort the curves by power
+            bids[t][gen] = dict(sorted(bids[t][gen].items()))
+
+            # make sure the curve is nondecreasing
+            pre_power = pmin
+            for power, marginal_cost in bids[t][gen].items():
+
+                # ignore pmin, because min load cost is special
+                if pre_power == pmin:
+                    pre_power = power
+                    continue
+                bids[t][gen][power] = max(bids[t][gen][power], bids[t][gen][pre_power])
+                pre_power = power
+
+            # calculate the actual cost
+            pre_power = 0
+            pre_cost = 0
+            for power, marginal_cost in bids[t][gen].items():
+
+                delta_p = power - pre_power
+                bids[t][gen][power] = pre_cost + marginal_cost * delta_p
+                pre_power = power
+                pre_cost += marginal_cost * delta_p
+
+        # check if bids are convex
+        for t in bids:
+            for gen in bids[t]:
+                if not self._is_convex_bid(bids[t][gen]):
+                    raise RuntimeError(
+                        f"Bids for generator {gen} at hour {t} is not convex!"
+                    )
+
+        return bids
+
+    # note: need to check form of input. it will return true with bad inputs
+    @staticmethod
+    def _is_convex_bid(bids):
```

Comment:
There's some code in Egret (which this implicitly depends on) which does this as well. https://github.com/grid-parity-exchange/Egret/blob/42eb712299cb32ee1e65e97069ee9b4320f49f3d/egret/model_library/transmission/tx_utils.py#L611.

You might try using that instead of implementing your own. This file demonstrates its use: https://github.com/grid-parity-exchange/Egret/blob/main/egret/model_library/transmission/tests/test_tx_utils.py